uploader/index_test.go

```diff
@@ -0,0 +1,611 @@
+	}
+	r.level, err = rdr.ReadUint32()
+	if err != nil {
+		return err
+	}
+	r.name, err = rdr.ReadString()
+	if err != nil {
+		return err
+	}
+	r.version, err = rdr.ReadUint32()
+
+	return err
+}
+
+func TestIndexParseFileDedup(t *testing.T) {
+	now32 := uint32(time.Now().Unix())
+	points := []point{
+		{
+			path:    "carbon.agents.carbon-clickhouse.writer.writtenBytes",
+			value:   1.0,
+			time:    1559465760,
+			date:    18049,
+			version: 1559465760,
+		},
+		{
+			// duplicate for index
+			path:    "carbon.agents.carbon-clickhouse.writer.writtenBytes",
+			value:   2.0,
+			time:    1559465800,
+			date:    18049,
+			version: 1559465800,
+		},
+		{
+			// duplicate parts for index
+			path:    "carbon.agents.carbon-clickhouse.tcp.receiver.errors",
+			value:   2.0,
+			time:    1559465800,
+			date:    18049,
+			version: 1559465800,
+		},
+		{
+			// next day
+			path:    "carbon.agents.carbon-clickhouse.tcp.receiver.errors",
+			value:   1.0,
+			time:    1662098177,
+			date:    19237,
+			version: 1662098177,
+		},
+		{
+			// tagged metric
+			path:    "errors?app=carbon-clickhouse&scope=tcp",
+			value:   1.0,
+			time:    1662098177,
+			date:    19237,
+			version: 1662098177,
+		},
+	}
+
+	wantIndexRecords := []indexRecord{
+		// carbon.agents.carbon-clickhouse.writer.writtenBytes
+		{date: 42, level: TreeLevelOffset + 5, name: "carbon.agents.carbon-clickhouse.writer.writtenBytes", version: now32},
+		{date: 42, level: TreeLevelOffset + 4, name: "carbon.agents.carbon-clickhouse.writer.", version: now32},
+		{date: 42, level: TreeLevelOffset + 3, name: "carbon.agents.carbon-clickhouse.", version: now32},
+		{date: 42, level: TreeLevelOffset + 2, name: "carbon.agents.", version: now32},
+		{date: 42, level: TreeLevelOffset + 1, name: "carbon.", version: now32},
+		{date: 42, level: ReverseTreeLevelOffset + 5, name: "writtenBytes.writer.carbon-clickhouse.agents.carbon", version: now32},
+		{date: 18049, level: 5, name: "carbon.agents.carbon-clickhouse.writer.writtenBytes", version: now32},
+		{date: 18049, level: ReverseLevelOffset + 5, name: "writtenBytes.writer.carbon-clickhouse.agents.carbon", version: now32},
+		// carbon.agents.carbon-clickhouse.tcp.errors
+		{date: 42, level: TreeLevelOffset + 6, name: "carbon.agents.carbon-clickhouse.tcp.receiver.errors", version: now32},
+		{date: 42, level: TreeLevelOffset + 5, name: "carbon.agents.carbon-clickhouse.tcp.receiver.", version: now32},
+		{date: 42, level: TreeLevelOffset + 4, name: "carbon.agents.carbon-clickhouse.tcp.", version: now32},
+		{date: 42, level: ReverseTreeLevelOffset + 6, name: "errors.receiver.tcp.carbon-clickhouse.agents.carbon", version: now32},
+		{date: 18049, level: 6, name: "carbon.agents.carbon-clickhouse.tcp.receiver.errors", version: now32},
+		{date: 18049, level: ReverseLevelOffset + 6, name: "errors.receiver.tcp.carbon-clickhouse.agents.carbon", version: now32},
+		// next day
+		// TODO (msaf1980): maybe can deduplicated, but not at now
```

Comment:
May be prevent write of this (at new day start) ? Need a write to cache not daily key only, but also with date: 42. It will be require 2x memory for cache. So not too simple.